### ISSUE TITLE
DB schema support re-broken

### ISSUE THREAD

**Issue by samroberton** [yogthos/migratus#90]
PR #70 implemented some changes to allow the user to specify a specific DB schema in the connection.

Issue #79 then reported that this caused `migratus` to attempt to create the `schema_migrations` table even though it already existed.  That issue was fixed by making the logic for detecting the existence of the `schema_migrations` table more promiscuous:  first, check for the `schema_migrations` table specifying the current schema, but if you don't find it, also check without specifying a schema.

That fix breaks schema support:  we do specify a schema, and we're seeing our CI fail like so:

```
org.postgresql.util.PSQLException: ERROR: relation "schema_migrations" does not exist
	  Position: 16
		at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2284)
		at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2003)
		at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:200)
		at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:424)
		at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:161)
		at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:114)
		at clojure.java.jdbc$db_query_with_resultset$run_query_with_params__3176.invoke(jdbc.clj:861)
		at clojure.java.jdbc$db_query_with_resultset.invokeStatic(jdbc.clj:877)
		at clojure.java.jdbc$db_query_with_resultset.invoke(jdbc.clj:847)
		at clojure.java.jdbc$query.invokeStatic(jdbc.clj:915)
		at clojure.java.jdbc$query.invoke(jdbc.clj:884)
		at clojure.java.jdbc$query.invokeStatic(jdbc.clj:898)
		at clojure.java.jdbc$query.invoke(jdbc.clj:884)
		at migratus.database$completed_ids_STAR_$fn__26375.invoke(database.clj:234)
```

I think this is happening because we have multiple schemas in our database, with the database shared by multiple environments each of which are using migratus.  I think `find-table` is finding a `schema_migrations` table in another schema, and assuming that it doesn't need to create one itself.

The underlying problem seems to be that the change to have `migratus` search for `schema_migrations` first in the given schema, and then search again without specifying a schema, makes the "does the table exist" check inconsistent with the ability to query/insert into the table.

To fix this properly, would it be reasonable to check for the existence of the table with a `SELECT 1 FROM schema_migrations`?  If that query then fails, then we'd create the table.

**Comment by yogthos** [comment]
That would probably be a good way to go. Another option might be to allow specifying the schema explicitly in the options map. I'd be open to either option.

**Comment by samroberton** [comment]
TL;DR:  PR incoming, but be aware that this PR does potentially make it more difficult to combine all migrations into a single transaction, if that's a future goal (and I think it should be).

The more long-winded version...

I don't particularly like the idea of specifying a schema explicitly in the options map.  My feeling is that `migratus` already requires more config than is strictly its concern, by assuming that it should always manage its own connection/disconnection based on a map of connection properties, rather than allowing for the possibility that maybe something other code has already established a pre-configured database connection that `migratus` could use.

Personally, I'd prefer to be able to give `migratus` an already-open connection, rather than a whole `db-spec` map.  In fact, I'd prefer to be able to give it an already-open transaction, so that all my migrations roll back if one of them fails.  If I have a production deployment that involves multiple migrations (relating to different features that happen to get deployed at the same time), and one of them fails, I'm going to back out the entire deployment -- I can't back out just those bits of code changes that relate to whichever migrations didn't successfully complete.  So it's currently a pain point that in that situation, `migratus` will leave the database partially migrated.

In any case, adding the schema to the `config` would mean that it has to be present there in two places:  once in the `db` map which is used to open the connection -- and there it's specified in a non-portable, JDBC-driver-specific way -- and then once in a common config item for `migratus` itself to understand.

With all that said, my dream of being able to pass in an open transaction also suffers from a conflict with my suggestion above:  if `SELECT 1 FROM schema_migrations` fails because the table doesn't exist yet, then on Postgres, that error causes the transaction to be marked rollback only, rendering it unusable.

So for now, I'm going to propose a PR which does a `SELECT 1 FROM schema_migrations` on a separate connection.  Some other time I'll try to give a bit more thought to how it might be possible to refactor things a little to allow the caller to have a little more control over how transactions are managed.

**Comment by yogthos** [comment]
I agree that passing an existing connection would be useful. Maybe the config map could have a `:connection` key that would be used instead of the `:db` key, and Migratus would use the supplied connection in that  case.

The `select 1` approach seems like the most reliable option, so I'm ok with the PR there. It does look like it fails travis though, so if you can fix it up, I can push out a new release with the change.

**Comment by samroberton** [comment]
Hadn't noticed that build failure -- thanks for mentioning it.  It's green now.

**Comment by yogthos** [comment]
Just pushed out `0.8.32` to Clojars with the fix, thanks for taking the lead on this.

**Comment by samroberton** [comment]
Thank you!
